networking/eth2/src/main/java/tech/pegasys/teku/networking/eth2/rpc/beaconchain/methods/BlobSidecarsByRootMessageHandler.java

```diff
@@ -0,0 +1,218 @@
+                          String.format(
+                              "Blob sidecar for blob identifier (%s) was not available",
+                              identifier));
+                    }
+                    return callback.respond(maybeSidecar.get());
+                  });
+    }
+
+    future.finish(callback::completeSuccessfully, err -> handleError(callback, err));
+  }
+
+  // MAX_REQUEST_BLOB_SIDECARS * MAX_BLOBS_PER_BLOCK
+  private int calculateMaxRequestSize() {
+    final UInt64 currentEpoch = combinedChainDataClient.getCurrentEpoch();
+    final int maxBlobsPerBlock =
+        SpecConfigDeneb.required(spec.atEpoch(currentEpoch).getConfig()).getMaxBlobsPerBlock();
+    return MAX_REQUEST_BLOB_SIDECARS.times(maxBlobsPerBlock).intValue();
+  }
+
+  private UInt64 getFinalizedEpoch() {
+    return combinedChainDataClient
+        .getFinalizedBlock()
+        .map(SignedBeaconBlock::getSlot)
+        .map(spec::computeEpochAtSlot)
+        .orElse(UInt64.ZERO);
+  }
+
+  /**
+   * Validations:
+   *
+   * <ul>
+   *   <li>A block for the block root is available
+   *   <li>The block root references a block greater than or equal to the minimum_request_epoch
+   * </ul>
+   */
+  private SafeFuture<Void> validateRequestedBlockRoot(
+      final Bytes32 blockRoot, final UInt64 finalizedEpoch) {
+    return combinedChainDataClient
+        .getBlockByBlockRoot(blockRoot)
+        .thenAcceptChecked(
+            maybeBlock -> {
+              if (maybeBlock.isEmpty()) {
+                throw new ResourceUnavailableException(
+                    String.format("Block for block root (%s) couldn't be retrieved", blockRoot));
+              }
+              final SignedBeaconBlock block = maybeBlock.get();
+              final UInt64 requestedEpoch = spec.computeEpochAtSlot(block.getSlot());
+              final UInt64 minimumRequestEpoch = computeMinimumRequestEpoch(finalizedEpoch);
+              if (requestedEpoch.isLessThan(minimumRequestEpoch)) {
+                throw new RpcException(
+                    INVALID_REQUEST_CODE,
+                    String.format(
+                        "Block root (%s) references a block earlier than the minimum_request_epoch (%s)",
+                        blockRoot, minimumRequestEpoch));
+              }
+            });
+  }
+
+  private UInt64 computeMinimumRequestEpoch(final UInt64 finalizedEpoch) {
+    final UInt64 currentEpoch = combinedChainDataClient.getCurrentEpoch();
+    return finalizedEpoch
+        .max(currentEpoch.minusMinZero(MIN_EPOCHS_FOR_BLOBS_SIDECARS_REQUESTS))
+        .max(denebForkEpoch);
+  }
+
+  private SafeFuture<Optional<BlobSidecar>> retrieveBlobSidecar(final BlobIdentifier identifier) {
+    return combinedChainDataClient.getBlobSidecarByBlockRootAndIndex(
+        identifier.getBlockRoot(), identifier.getIndex());
+  }
+
+  private void handleError(final ResponseCallback<BlobSidecar> callback, final Throwable error) {
```

Comment:
It looks very generic. How difficult is to move it to the parent class?

Reply:
Not very much but there are logs which are different and not straightforward to parameterize so thinking of leaving it for now